### PR TITLE
Issues/#7638 configure alerts from am

### PR DESCRIPTION
**Pull request by farmborough**
See https://github.com/gravitee-io/gravitee-access-management/pull/1928

## :id: 
Fixes https://github.com/gravitee-io/issues/issues/7638

## :pencil2: 

Please disregards commit 8f385cb3ed26ba85f2e676511969f04c4cc444d4 as it has been cherry picked from @remisultan's corresponding [PR](feat/risk-assessment-feature).

Two main changes in this PR:

1. Additional Risk Assessment alert that can be enabled/disabled in the UI (see screenshot below)
2. New class in `RiskAssessmentHelper`, used by `AlertEventProcessor` for retrieving these configured values which are then used to compute the risk assessment at the risk assessment service.

## :memo: Test scenarios 

In the absence of the risk assessment service the most that can be manually tested is that enabling/disabling the risk assessment alert in AM leads to an alert trigger being created/removed in the Alert Engine. An instance of Alert Engine is therefore required as well as the AM plugin for connecting to Alert Engine.

## :computer: Add screenshots for UI

Screenshot 2022-06-20 at 09 03 54

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java [hunk 1]
```diff
@@ -86,21 +98,38 @@ public void handle(RoutingContext routingContext) {
...
+            AssessmentSettings assessmentSettings = new AssessmentSettings().setEnabled(enabled);
+            if(enabled) {
+                assessmentSettings.setThresholds((Map<Assessment, Double>) thresholds);
+                settings.setEnabled(true);
+            }
+            setter.accept(settings, assessmentSettings);
+        });
+        return settings;
+    }
+
```
Comment (remisultan) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java:133]: I am not very certain that the Alert risk Assessment has to be processed here.

`io.gravitee.am.gateway.handler.common.alert.AlertEventProcessor` contains everything necessary to process a user authentication event.

If the authentication successful, you can check it within the AlertEventProcessor
```
                .property(PROPERTY_AUTHENTICATION_STATUS, eventItem.type().name())
```

then we decorate the AlertEvent with the assessment message

Maybe there is some abstraction work to think about so all the utility methods in risk assessment are reusable

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml [hunk 2]
```diff
@@ -354,3 +366,45 @@ ds:
...
+        name: "Too many login failures detected"
+        description: "More than {threshold}% of logins are in failure over the last {window} second(s)"
+        # the threshold rate in % to reach before notify. Default 10% of login failures.
+        threshold: 1
+        # the minimum sample size. Default 1000 login attempts.
+        sampleSize: 1
+        # window time in seconds. Default 600s (10 minutes).
+        window: 10
+        # severity of the alert (INFO, WARNING, CRITICAL). Default WARNING.
+        severity: WARNING
```
Comment (remisultan) [inline gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:411]: We may have to document these but we should not have them directly in the `gravitee.yml`

File: gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml [hunk 3]
```diff
@@ -319,6 +319,18 @@ user:
       #subject: Please reset your password
     token:
     #expire-after: 300 # the end of validity of the token in seconds for reset password email (default 300 = 5 minutes)
+  activity:
```
Comment (remisultan) [inline gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml:322]: We may have to document these but we should not have this configuration directly in the gravitee.yml

File: gravitee-am-ui/src/app/domain/alerts/general/general.component.ts [hunk 4]
```diff
@@ -106,6 +106,12 @@ export class DomainAlertGeneralComponent implements OnInit {
         icon: 'account_box',
         available: true
       },
+      risk_assessment: {
+        name: 'Risk assessment',
+        description: ' Triggers a risk assessment',
```
Comment (remisultan) [inline gravitee-am-ui/src/app/domain/alerts/general/general.component.ts:111]: Forward space
```ts
        description: 'Triggers a risk assessment',
```

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/alerts/RiskAssessmentAlert.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.core.env.Environment;
+
+import static io.gravitee.am.common.event.AlertEventKeys.PROPERTY_DOMAIN;
+import static io.gravitee.am.management.service.alerts.AlertTriggerFactory.AUTHENTICATION_SOURCE;
+import static java.util.Collections.singletonList;
+
+/**
+ * Risk assessment alert.
+ *
+ * @author GraviteeSource Team
```
Comment (remisultan) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/alerts/RiskAssessmentAlert.java:31]: Author is missing 😄

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java [hunk 6]
```diff
@@ -86,21 +98,38 @@ public void handle(RoutingContext routingContext) {
...
     private void computeRiskAssessment(RoutingContext routingContext, Client client, io.gravitee.am.model.User user, RiskAssessmentSettings riskAssessment) {
         var assessmentMessage = Single.just(new AssessmentMessage().setSettings(riskAssessment).setData(new AssessmentData()));
         assessmentMessage
                 .flatMap(buildDeviceAssessmentMessage(routingContext, riskAssessment, client, user))
                 .flatMap(buildIpReputationMessage(routingContext, riskAssessment))
                 .flatMap(buildGeoVelocityMessage(client, user, riskAssessment))
-                .doOnSuccess(message -> decorateWithRiskAssessment(routingContext, message))
+                .doOnSuccess(message ->     decorateWithRiskAssessment(routingContext, message))
```
Comment (remisultan) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java]: Forward space

**Review by remisultan — CHANGES_REQUESTED** [review]
The Risk Assessment Handler in the OAuth2Provider and the Risk Asssessment processed for Alerts should be distinct

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml (hunk 2 above)
Comment (farmborough) [reply gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:411]: Which ones?

Comment (farmborough) [reply gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:411]: Ah I understand.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java (hunk 1 above)
Comment (farmborough) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java:133]: Yes I would have thought at least `RiskAssessmentHandler#computeRiskAssessment` should be available to `AlertEventProcessor` as well.

Comment (farmborough) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java:133]: Which in turn depends on three other methods and even more fields (e..g EventBus).

**Comment by farmborough** [comment]
Now on https://github.com/gravitee-io/gravitee-access-management/pull/1928

**Pull request closed without merging**